web/source/kmwcallback.ts

```diff
@@ -234,6 +324,143 @@ class KeyboardInterface {
+
+        var unmatchedDeadkeys = [].concat(this._DeadKeys);
+
+        // Time to build from scratch!
+        var index = 0;
+        cache = { valContext: [], deadContext: []};
+        while(cache.valContext.length < n) {
+          // As adapted from `deadkeyMatch`.
+          var sp=this._SelPos(Ptarg);
+          var deadPos = sp - index;
+          if(unmatchedDeadkeys.length > 0 && unmatchedDeadkeys[0].p == deadPos) {
+            // Take the deadkey.
+            cache.deadContext[n-cache.valContext.length-1] = unmatchedDeadkeys[0];
+            cache.valContext = [unmatchedDeadkeys[0].d].concat(cache.valContext);
+            unmatchedDeadkeys.splice(0, 1);
+          } else {
+            // Take the character.  We get "\ufffe" if it doesn't exist.
+            var kc = this.context(++index, 1, Ptarg);
+            cache.valContext = ([kc] as (string|number)[]).concat(cache.valContext);
+          }
+        }
+        this.cachedContextEx.set(n, n, cache);
+      }
+
+      // Now that we have the cache...
+      var subCache = cache;
+      subCache.valContext = subCache.valContext.slice(0, ln);
+      for(var i=0; i < subCache.valContext.length; i++) {
+        if(subCache[i] == '\ufffe') {
+          subCache.valContext.splice(0, 1);
+          subCache.deadContext.splice(0, 1);
+        }
+      }
+
+      if(subCache.valContext.length == 0) {
+        subCache.valContext = ['\ufffe'];
+        subCache.deadContext = [];
+      }
+
+      this.cachedContextEx.set(n, ln, subCache);
+
+      return subCache;
+    }
+  }
+
+  /**
+   * Function       fullContextMatch    KFCM
+   * Scope          Private
+   * @param         {number}    n       Number of characters to move back from caret
+   * @param         {Object}    Ptarg   Focused element
+   * @param         {Array}     rule    An array of ContextEntries to match.
+   * @return        {boolean}           True if the fully-specified rule context matches the current KMW state.
+   * 
+   * A KMW 10+ function designed to bring KMW closer to Keyman Desktop functionality,
+   * near-directly modeling (externally) the compiled form of Desktop rules' context section.
+   */
+  fullContextMatch(n: number, Ptarg: HTMLElement, rule: ContextEntry[]): boolean {
+    // Stage one:  build the context index map.
+    var fullContext = this._BuildExtendedContext(n, rule.length, Ptarg);
+    var context = fullContext.valContext;
+    var deadContext = fullContext.deadContext;
+
+    var mismatch = false;
+
+    // Stage two:  time to match against the rule specified.
+    for(var i=0; i < rule.length; i++) {
+      if(typeof(rule[i]) == 'string') {
+        var str = rule[i] as string;
+        if(str != context[i]) {
+          mismatch = true;
+          break;
+        }
+      } else if(rule[i]['d'] !== undefined) {
+        var deadSpec = rule[i] as RuleDeadkey;
+        // We still need to set a flag here; 
+        if(deadSpec['d'] != context[i]) {
+          mismatch = true;
```

Comment:
Should we be `break`ing after any mismatch? Or could we refactor the loop into another function and short-circuit `return false on mismatch`? It'd simplify the code a little bit.